orca/service.go

```diff
@@ -0,0 +1,214 @@
+	// type assertion any more, since the new versions accept a
+	// grpc.ServiceRegistrar in their registration functions.
+	srv, ok := s.(*grpc.Server)
+	if !ok {
+		return nil, fmt.Errorf("concrete type of provided grpc.ServiceRegistrar is %T, only supported type is %T", s, &grpc.Server{})
+	}
+
+	// The default minimum supported reporting interval value can be overridden
+	// for testing purposes through the orca internal package.
+	if !opts.allowAnyMinReportingInterval {
+		if opts.MinReportingInterval < 0 || opts.MinReportingInterval < minReportingInterval {
+			opts.MinReportingInterval = minReportingInterval
+		}
+	}
+	service := &Service{
+		minReportingInterval: opts.MinReportingInterval,
+		utilization:          make(map[string]float64),
+	}
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(srv, service)
+	return service, nil
+}
+
+// determineReportingInterval determines the reporting interval for out-of-band
+// metrics. If the reporting interval is not specified in the request, or is
+// negative or is less than the configured minimum (via
+// ServiceOptions.MinReportingInterval), the latter is used. Else the value from
+// the incoming request is used.
+func (s *Service) determineReportingInterval(req *v3orcaservicegrpc.OrcaLoadReportRequest) time.Duration {
+	if req.GetReportInterval() == nil {
+		return s.minReportingInterval
+	}
+	dur := req.GetReportInterval().AsDuration()
+	if dur < 0 || dur < s.minReportingInterval {
+		logger.Warningf("Received reporting interval %q is less than configured minimum: %v. Using default: %s", dur, s.minReportingInterval)
+		return s.minReportingInterval
+	}
+	return dur
+}
+
+func (s *Service) sendMetricsResponse(stream v3orcaservicegrpc.OpenRcaService_StreamCoreMetricsServer) error {
+	return stream.Send(s.toLoadReportProto())
+}
+
+// StreamCoreMetrics streams custom backend metrics injected by the server
+// application.
+func (s *Service) StreamCoreMetrics(req *v3orcaservicepb.OrcaLoadReportRequest, stream v3orcaservicepb.OpenRcaService_StreamCoreMetricsServer) error {
+	// Send one response immediately before creating and waiting for the timer
+	// to fire.
+	if err := s.sendMetricsResponse(stream); err != nil {
+		return err
+	}
+
+	ticker := time.NewTicker(s.determineReportingInterval(req))
+	defer ticker.Stop()
+
+	for {
+		// Send a response containing the currently recorded metrics
+		select {
+		case <-stream.Context().Done():
+			return status.Error(codes.Canceled, "Stream has ended.")
+		case <-ticker.C:
+			if err := s.sendMetricsResponse(stream); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// SetCPUUtilization records a measurement for the CPU utilization metric.
+func (s *Service) SetCPUUtilization(val float64) {
+	s.mu.Lock()
+	s.cpu = val
+	s.mu.Unlock()
+}
+
+// DeleteCPUUtilization deletes a previously recorded measurement for the CPU
+// utilization metric.
+func (s *Service) DeleteCPUUtilization() {
+	s.mu.Lock()
+	s.cpu = 0
```

Comment:
Is there no difference between zero CPU utilization and deleted (presumably "not reporting")?  If not, do we need this API?

Reply:
cpu, and memory are special because they have dedicated fields in the orca load report proto. I initially had the delete API for them, just to be in sync with Java. But I think it does not make much sense to have these delete APIs for two reasons: 
1. they can always explicitly set it to `0`. 
2. actual applications which are setting cpu and memory will hardly have a reason to delete them.

We can always add it if there is demand for it.

The other issue of not reporting when they are not set is not something that I'm doing right now. Maybe it will save a few bytes if done. Do you think it is worth doing?

Reply:
> The other issue of not reporting when they are not set is not something that I'm doing right now. Maybe it will save a few bytes if done. Do you think it is worth doing?

I'm unsure of the semantic differences between "unset" and "set to zero".  If the protocol supports both (optional field vs. always present) then do we need to model that difference in our API as well?  Did Java?